app/controllers/application_controller.rb

```diff
@@ -1,10 +1,169 @@
 require './config/environment'
 
 class ApplicationController < Sinatra::Base
-
   configure do
     set :public_folder, 'public'
     set :views, 'app/views'
+    enable :sessions
+    set :session_secret, "secret"
+  end
+
+get '/' do
+  if session[:user_id]
+     redirect '/tweets'
+   end
+  erb :index
+
+end
+
+get '/signup' do
+if session[:user_id]
+   redirect '/tweets'
+ end
+  erb :'users/create_user'
+end
+
+get '/login' do
+  if session[:user_id]
+     redirect '/tweets'
+   end
+  erb :'users/login'
+end
+
+
+
+post '/login' do
+  @user = User.find_by(:username => params[:username])
+  if @user && @user.authenticate(params[:password])
+  session[:user_id]= @user.id
+  redirect '/tweets'
+else
+  redirect '/login'
+end
+end
+
+
+get '/users/:id' do
+  @current_user = User.find_by(:username => params[:captures])
+  erb :'/users/show'
+end
+
+get '/tweets/new' do
+if  !is_logged_in?
+  redirect '/login'
+else
+  erb :'tweets/create_tweet'
+end
+end
+
+get '/tweets/:id' do
+  if  !is_logged_in?
+    redirect '/login'
+  else
+    @tweet = Tweet.find_by_id(params[:id])
+    erb :'/tweets/show_tweet'
+  end
+end
+
+get '/tweets/:id/edit' do
+  if  !is_logged_in?
+    redirect '/login'
+  else
+    @tweet = Tweet.find_by_id(params[:id])
+    erb :'/tweets/edit_tweet'
+  end
+end
+
+
+patch '/tweets/:id' do
+if
+  !is_logged_in?
+    redirect '/login'
+
+else
+  @tweet = Tweet.find(params[:id])
+    if @tweet.content.empty?
+    redirect "/tweets/#{@tweet.id}/edit"
+  end
+  @tweet.update(content: params[:content])
+  @tweet.save
+  redirect "/tweets/#{@tweet.id}/edit"
+
+end
+end
+
+
+
+
+
+
+post '/tweets/new' do
+  is_logged_in?
+
+  if params[:content].empty?
+    redirect '/tweets/new'
+  else @tweet = Tweet.create(params)
+  end
+  @current_user.tweets << @tweet
+
+  redirect '/tweets'
+end
+
+
+post '/signup' do
+
+  if params[:username].empty?
+     redirect '/signup'
+
+    elsif params[:email].empty?
+      redirect '/signup'
+
+    elsif params[:password].empty?
+      redirect '/signup'
+
+    else @current_user = User.create(params)
+      session[:user_id]= @current_user.id
+      redirect '/tweets'
+
+    end
+end
+
+get '/logout' do
+  if is_logged_in?
+  session.clear
+  redirect '/login'
+else
+  redirect '/'
+end
+end
+
+delete '/tweets/:id/delete' do
+if
+    !is_logged_in?
+    redirect '/login'
+else
+  @tweet = Tweet.find_by_id(params[:id])
```

Comment:
Prefer to find tweets that are already only associated to the `current_user` than to manually check the `user_id` foreign key of the tweet against the `current_user.id`

```ruby
tweet = current_user.tweets.find_by(id: params[:id])
if tweet && tweet.destroy
  redirect "/tweets"
else
  redirect "/tweets/#{tweet.id}"
end
```

We're using the association collection method provided by [`has_many`](http://api.rubyonrails.org/classes/ActiveRecord/Associations/ClassMethods.html#method-i-has_many) to query the user's tweets. 

Note the usage of [`find_by`](http://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-find-by) vs [`find`](http://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-find). Had we used `find` here, ActiveRecord would raise an error in the event that we could not find a tweet associated with this user. `find_by` will allow this action to fail gracefully, simply returning nil if the user is trying to delete a tweet that they do not have access to.